sunpy/map/mapbase.py

```diff
@@ -1482,7 +1484,14 @@ def rotate(self, angle: u.deg = None, rmatrix=None, order=4, scale=1.0,
         pad_x = int(np.max((diff[1], 0)))
         pad_y = int(np.max((diff[0], 0)))
 
-        new_data = np.pad(self.data,
+        if issubclass(self.data.dtype.type, numbers.Integral):
```

Comment:
I do not think that always casting to float64 is a good idea. People often want to use rotate-like functionality to rotate many maps, and if you take a unit8 image as input here and cast it up to float64 you are multiplying the memory usage by a factor of 8! I know that if scikit-image is installed by default we cast up anyway, but with efforts like #4502 I really don't think doing this in an implementation-agnostic way is a good idea.

Why don't we raise a warning if missing is set to nan and the dtype is not float?

Reply:
I guess so, do we raise a warning and then proceed to cast it when missing=np.nan or just raise a warning and let the erroneous plot show up?

Reply:
The other option is to raise an exception if the missing value doesn't have the same dtype (or a compatible one) with the data array? I am not sure if that's a good idea though.

Reply:
The plots do say the "bad values (NaN) are given in red". If nan is considered bad then I guess raising an warning and converting to 0 is better?

Reply:
Another option is we throw a warning and then cast the data array to float64 if the missing value is an incompatible dtype with the array.

It basically comes down to which dtype do we want to prioritise. If the dtype of the array is the most important then we should probably error if missing is incompatible with the array dtype (which is potentially problematic if the default causes an exception). If we want to prioritise the dtype of missing then we should cast the array to a compatible dtype if it is incompatible with the missing value.

I think I need some thoughts from other people on this.

Reply:
Do we have no control over the missing value?

Reply:
I suppose, I shall leave this PR open for that then, thanks!

Reply:
@nabobalis depending on how it's being cast, if passing np.nan (float) to an int map causes this issue.